### ISSUE TITLE
Add support for PostgreSQL 16/17

### ISSUE THREAD

**Issue by Vonng** [tvondra/lower_quantile#1]
Got compile errors when building against PostgreSQL 16/17

```
$ rpmbuild --with debuginfo --define "pgmajorversion 16" -ba ~/rpmbuild/SPECS/lower_quantile.spec
Executing(%prep): /bin/sh -e /var/tmp/rpm-tmp.DXDkau
+ umask 022
+ cd /home/vagrant/rpmbuild/BUILD
+ cd /home/vagrant/rpmbuild/BUILD
+ rm -rf lower_quantile-1.0.0
+ /usr/bin/gzip -dc /home/vagrant/rpmbuild/SOURCES/lower_quantile-1.0.0.tar.gz
+ /usr/bin/tar -xof -
+ STATUS=0
+ '[' 0 -ne 0 ']'
+ cd lower_quantile-1.0.0
+ /usr/bin/chmod -Rf a+rX,u+w,g-w,o-w .
+ exit 0
Executing(%build): /bin/sh -e /var/tmp/rpm-tmp.h3M1Yb
+ umask 022
+ cd /home/vagrant/rpmbuild/BUILD
+ cd lower_quantile-1.0.0
+ PATH=/usr/pgsql-16/bin:/home/vagrant/.cargo/bin:/home/vagrant/.local/bin:/home/vagrant/bin:/usr/share/Modules/bin:/usr/lib64/ccache:/usr/local/bin:/usr/bin:/usr/local/sbin:/usr/sbin
+ /usr/bin/make -j10
gcc -Wall -Wmissing-prototypes -Wpointer-arith -Wdeclaration-after-statement -Werror=vla -Wendif-labels -Wmissing-format-attribute -Wimplicit-fallthrough=3 -Wcast-function-type -Wshadow=compatible-local -Wformat-security -fno-strict-aliasing -fwrapv -fexcess-precision=standard -Wno-format-truncation -Wno-stringop-truncation -O2 -g -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -fexceptions -fstack-protector-strong -grecord-gcc-switches -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1 -m64 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -fPIC -fvisibility=hidden -fPIC -fvisibility=hidden -I. -I./ -I/usr/pgsql-16/include/server -I/usr/pgsql-16/include/internal  -D_GNU_SOURCE -I/usr/include/libxml2  -I/usr/include  -c -o lower_quantile.o lower_quantile.c
/usr/bin/clang -Wno-ignored-attributes -fno-strict-aliasing -fwrapv -fexcess-precision=standard -Wno-unused-command-line-argument -Wno-compound-token-split-by-macro -O2  -I. -I./ -I/usr/pgsql-16/include/server -I/usr/pgsql-16/include/internal  -D_GNU_SOURCE -I/usr/include/libxml2  -I/usr/include -flto=thin -emit-llvm -c -o lower_quantile.bc lower_quantile.c
lower_quantile.c:565:28: error: incompatible integer to pointer conversion passing 'Datum' (aka 'unsigned long') to parameter of type 'Numeric' (aka 'struct NumericData *') [-Wint-conversion]
  565 |                                                                                         NumericGetDatum(keys[i])));
      |                                                                                                         ^~~~~~~
/usr/pgsql-16/include/server/fmgr.h:643:44: note: expanded from macro 'DirectFunctionCall1'
  643 |         DirectFunctionCall1Coll(func, InvalidOid, arg1)
      |                                                   ^~~~
/usr/pgsql-16/include/server/utils/numeric.h:72:25: note: passing argument to parameter 'X' here
   72 | NumericGetDatum(Numeric X)
      |                         ^
lower_quantile.c:592:24: error: incompatible integer to pointer conversion passing 'Datum' (aka 'unsigned long') to parameter of type 'const void *' [-Wint-conversion]
  592 |         PG_RETURN_ARRAYTYPE_P(makeArrayResult(astate,
      |                               ^~~~~~~~~~~~~~~~~~~~~~~
  593 |                                                                                   CurrentMemoryContext));
      |                                                                                   ~~~~~~~~~~~~~~~~~~~~~
/usr/pgsql-16/include/server/utils/array.h:265:54: note: expanded from macro 'PG_RETURN_ARRAYTYPE_P'
  265 | #define PG_RETURN_ARRAYTYPE_P(x)          PG_RETURN_POINTER(x)
      |                                                             ^
/usr/pgsql-16/include/server/fmgr.h:361:53: note: expanded from macro 'PG_RETURN_POINTER'
  361 | #define PG_RETURN_POINTER(x) return PointerGetDatum(x)
      |                                                     ^
/usr/pgsql-16/include/server/postgres.h:322:29: note: passing argument to parameter 'X' here
  322 | PointerGetDatum(const void *X)
      |                             ^
2 errors generated.
make: *** [/usr/pgsql-16/lib/pgxs/src/makefiles/../../src/Makefile.global:1093: lower_quantile.bc] Error 1
make: *** Waiting for unfinished jobs....
In file included from /usr/pgsql-16/include/server/utils/array.h:64,
                 from lower_quantile.c:15:
lower_quantile.c: In function 'numeric_array_to_double':
lower_quantile.c:565:32: warning: passing argument 1 of 'NumericGetDatum' makes pointer from integer without a cast [-Wint-conversion]
            NumericGetDatum(keys[i])));
                            ~~~~^~~
/usr/pgsql-16/include/server/fmgr.h:643:44: note: in definition of macro 'DirectFunctionCall1'
  DirectFunctionCall1Coll(func, InvalidOid, arg1)
                                            ^~~~
In file included from lower_quantile.c:17:
/usr/pgsql-16/include/server/utils/numeric.h:72:25: note: expected 'Numeric' {aka 'struct NumericData *'} but argument is of type 'Datum' {aka 'long unsigned int'}
 NumericGetDatum(Numeric X)
                 ~~~~~~~~^
In file included from /usr/pgsql-16/include/server/utils/array.h:64,
                 from lower_quantile.c:15:
lower_quantile.c: In function 'double_to_array':
lower_quantile.c:592:24: warning: passing argument 1 of 'PointerGetDatum' makes pointer from integer without a cast [-Wint-conversion]
  PG_RETURN_ARRAYTYPE_P(makeArrayResult(astate,
                        ^~~~~~~~~~~~~~~~~~~~~~~
             CurrentMemoryContext));
             ~~~~~~~~~~~~~~~~~~~~~
/usr/pgsql-16/include/server/fmgr.h:361:53: note: in definition of macro 'PG_RETURN_POINTER'
 #define PG_RETURN_POINTER(x) return PointerGetDatum(x)
                                                     ^
lower_quantile.c:592:2: note: in expansion of macro 'PG_RETURN_ARRAYTYPE_P'
  PG_RETURN_ARRAYTYPE_P(makeArrayResult(astate,
  ^~~~~~~~~~~~~~~~~~~~~
In file included from lower_quantile.c:14:
/usr/pgsql-16/include/server/postgres.h:322:29: note: expected 'const void *' but argument is of type 'Datum' {aka 'long unsigned int'}
 PointerGetDatum(const void *X)
                 ~~~~~~~~~~~~^
error: Bad exit status from /var/tmp/rpm-tmp.h3M1Yb (%build)


RPM build errors:
    Bad exit status from /var/tmp/rpm-tmp.h3M1Yb (%build)
```

**Comment by tvondra** [comment]
Should be fixed in 12c48a7dffc227b8d72c8501f4c705bf1f425ec3 and 65b607769a3983b607ea0515beac8bcae836c960. Can you try?

**Comment by Vonng** [comment]
> Should be fixed in [12c48a7](https://github.com/tvondra/lower_quantile/commit/12c48a7dffc227b8d72c8501f4c705bf1f425ec3) and [65b6077](https://github.com/tvondra/lower_quantile/commit/65b607769a3983b607ea0515beac8bcae836c960). Can you try?

Compiling from the HEAD has resolved the issue. Thanks for your response. 😄
